### PR TITLE
refactor (jkube-kit/config/image): Migrate all tests from JUnit4 to JUnit5 (#1552)

### LINKED ISSUE #1552: (jkube-kit-config-image): Replace JUnit4 with JUnit5
### Component

jkube-kit-config-image

### Task description

Child of #1546

- Remove JUnit4 dependencies
- Add JUnit5 dependencies
- Refactor tests

### PR DESCRIPTION
**Pull request by anurag-rajawat**
## Description

Fixes #1552 

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [x] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [x] My code follows the style guidelines of this project
 - [x] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [x] New and existing unit tests pass locally with my changes
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java [hunk 1]
```diff
@@ -179,42 +154,76 @@ public void namesUsedByDockerTests() {
...
+        assertThat(imageName.getUser()).isNull();
+        assertThat(imageName.getRegistry()).isEqualTo("foo.com:5000");
+        assertThat(imageName.getRepository()).isEqualTo("customer-service-cache");
+        assertThat(imageName.getTag()).isEqualTo("latest");
+    }
+
+    // TODO: Meaningful display name
+    @ParameterizedTest
+    @MethodSource("registryNamingTestData")
+    void testRegistryNaming(String imageName, String registry, String expected) {
```
Comment (anurag-rajawat) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:191]: @manusa please suggest a meaningful name for this test

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java [hunk 2]
```diff
@@ -179,42 +154,76 @@ public void namesUsedByDockerTests() {
...
+        assertThat(new ImageName(imageName).getFullName(registry)).isEqualTo(expected);
+    }
+
+    // TODO: Meaningful display name
+    @ParameterizedTest
+    @MethodSource("registryNamingExtendedTestData")
+    void testRegistryNamingExtended(String imageName, String registry, String expected) {
```
Comment (anurag-rajawat) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:198]: https://github.com/eclipse/jkube/pull/1659/files#r921838135

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java [hunk 3]
```diff
@@ -13,146 +13,95 @@
...
-                .mode(HealthCheckMode.cmd)
-                .cmd(Arguments.builder().shell("exit 0").build())
-                .retries(1)
-                .interval("2s")
-                .timeout("3s")
-                .startPeriod("30s")
-                .build()
-                .validate();
-    }
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java:32]: Why is this required? I'm not sure if tests are dependent on each other.

Comment (anurag-rajawat) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java:32]: These tests are not dependent on each other. If we don't add this order, `Good Health Check 7` and `Bad Health Check 7` are executed first and after that rest of the tests are executed which doesn't look good. So, executing these in this order, all `Good health check tests` are executed first and after that, `Bad health check tests` are executed.

Comment (rohanKanojia) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java:32]: I'm not sure if this is a good reason for making tests ordered.

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/CleanupModeTest.java [hunk 4]
```diff
@@ -41,20 +40,17 @@ public void parse() {
...
         try {
             CleanupMode.parse("blub");
             fail();
         } catch (IllegalArgumentException exp) {
-            assertTrue(exp.getMessage().contains("blub"));
-            assertTrue(exp.getMessage().contains("try"));
-            assertTrue(exp.getMessage().contains("none"));
-            assertTrue(exp.getMessage().contains("remove"));
+            assertThat(exp.getMessage()).contains("blub", "try", "none", "remove");
         }
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/CleanupModeTest.java:54]:
```java
        assertThatIllegalArgumentException()
            .isThrownBy(() -> CleanupMode.parse("blub"))
            .withMessageContainingAll("blub", "try", "none", "remove");
```

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilderTest.java [hunk 5]
```diff
@@ -212,63 +210,65 @@ public void testHealthCheckCmdParams() {
...
+              .doesNotContain("/export");
     }
 
     @Test
-    public void testDockerFileKeywords() {
+    void testDockerFileKeywords() {
         StringBuilder b = new StringBuilder();
         DockerFileKeyword.RUN.addTo(b, "apt-get", "update");
-        assertEquals("RUN apt-get update\n", b.toString());
+        assertThat(b.toString()).hasToString("RUN apt-get update\n");
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilderTest.java:263]:
```suggestion
        assertThat(b).hasToString("RUN apt-get update\n");
```

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilderTest.java [hunk 6]
```diff
@@ -212,63 +210,65 @@ public void testHealthCheckCmdParams() {
...
 
         b = new StringBuilder();
         DockerFileKeyword.EXPOSE.addTo(b, "1010", "2020");
-        assertEquals("EXPOSE 1010 2020\n",b.toString());
+        assertThat(b.toString()).hasToString("EXPOSE 1010 2020\n");
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilderTest.java:267]:
```suggestion
        assertThat(b).hasToString("EXPOSE 1010 2020\n");
```

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilderTest.java [hunk 7]
```diff
@@ -212,63 +210,65 @@ public void testHealthCheckCmdParams() {
...
 
         b = new StringBuilder();
         DockerFileKeyword.USER.addTo(b, "roland");
-        assertEquals("USER roland\n",b.toString());
+        assertThat(b.toString()).hasToString("USER roland\n");
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/DockerFileBuilderTest.java:271]:
```suggestion
        assertThat(b).hasToString("USER roland\n");
```

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java (hunk 3 above)
Comment (anurag-rajawat) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java:32]: I'll remove it

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java (hunk 1 above)
Comment (anurag-rajawat) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:191]: @rohanKanojia please suggest descriptive name for this test

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/CleanupModeTest.java [hunk 8]
```diff
@@ -41,20 +41,19 @@ public void parse() {
...
             CleanupMode.parse("blub");
             fail();
         } catch (IllegalArgumentException exp) {
-            assertTrue(exp.getMessage().contains("blub"));
-            assertTrue(exp.getMessage().contains("try"));
-            assertTrue(exp.getMessage().contains("none"));
-            assertTrue(exp.getMessage().contains("remove"));
+            assertThatIllegalArgumentException()
+                    .isThrownBy(() -> CleanupMode.parse("blub"))
+                    .withMessageContainingAll("blub", "try", "none", "remove");
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/CleanupModeTest.java:56]: I'm sorry for not being clear. I meant to replace whole try-catch block with this

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java (hunk 1 above)
Comment (rohanKanojia) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:192]: I think maybe we can try this
```suggestion
    @DisplayName("getFullName should add registry to image name when valid registry is provided")
    void getFullName_whenRegistryProvided_shouldReturnImageNameWithRegistry(String imageName, String registry, String expected) {
```

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java (hunk 2 above)
Comment (rohanKanojia) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java]: I'm slightly confused. This test is same as `testRegistryNamingExtended`. Why have you created it? Can't we test all these in `testRegistryNaming` ?

Comment (anurag-rajawat) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java]: Yes we can test all these in `testRegistryNaming` test, but I created it because it was there
https://github.com/eclipse/jkube/blob/c40de91d9604ee433f2403e87057078e4775644b/jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java#L132

**Comment by anurag-rajawat** [comment]
@rohanKanojia can we replace this whole try-catch block
https://github.com/eclipse/jkube/blob/c40de91d9604ee433f2403e87057078e4775644b/jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java#L166-L169

with this ?
```java
assertThatIllegalArgumentException()
                    .as("Name '%s' should fail", i)
                    .isThrownBy(() -> new ImageName(i));
```

**Comment by rohanKanojia** [comment]
@anurag-rajawat : Yeah, looks correct to me :+1:

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java (hunk 2 above)
Comment (rohanKanojia) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java]: I see. I think it's safe to merge both tests.

**Review by rohanKanojia — APPROVED** [review]

**Review by manusa — APPROVED** [review]

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java (hunk 3 above)
Comment (manusa) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/build/HealthCheckConfigTest.java]: The ideal option would have been to rename the tests, `Good Health Check 7` is a very good example of a bad name for a test :sweat_smile: 
I have no idea what these tests are testing, nor what the spec says about the code they are verifying.
